### ISSUE TITLE
how to get good weight

### ISSUE THREAD

**Issue by mjanddy** [zylo117/Yet-Another-EfficientDet-Pytorch#15]
Hi,if I want to get the weight as your,how I set the train parameter？

**Comment by zylo117** [comment]
Do you mean training on coco from scratch? It's pointless and boring to reinvent wheels, and there is almost no trick at all but tuning lr once in a while. It just takes weeks to converge if you're lucky enough, months if not, like me. I recommend using Adamw instead of SGD until the very final stage, admaw optimizes params way faster than SGD in this model.

**Comment by kaishijeng** [comment]
>>It just takes weeks to converge if you're lucky enough, months if not
Kind too long. I have trained detectron2 and it took a couple days to converge with one GPU.
I am training d0 with coco dataset now and will let it train for 40 epoch. Then check the quality.
Below is test result at 16th epoch and it seems something is wrong with bounding box. All bboxes are long rectangle: 
![image](https://user-images.githubusercontent.com/12689630/78970409-1309b680-7abe-11ea-9ac7-d2c7047ca95c.png)

**Comment by zylo117** [comment]
> > > It just takes weeks to converge if you're lucky enough, months if not
> > > Kind too long. I have trained detectron2 and it took a couple days to converge with one GPU.
> > > I am training d0 with coco dataset now and will let it train for 40 epoch. Then check the quality.
> > > Below is test result at 16th epoch and it seems something is wrong with bounding box. All bboxes are long rectangle:
> > > ![image](https://user-images.githubusercontent.com/12689630/78970409-1309b680-7abe-11ea-9ac7-d2c7047ca95c.png)

Did you change anchor config? Can you share your weight for debug propose?

But it's quite normal if you train from scratch, I got lot of error boxes/class in the early stage

**Comment by zylo117** [comment]
> > > It just takes weeks to converge if you're lucky enough, months if not
> > > Kind too long. I have trained detectron2 and it took a couple days to converge with one GPU.
> > > I am training d0 with coco dataset now and will let it train for 40 epoch. Then check the quality.
> > > Below is test result at 16th epoch and it seems something is wrong with bounding box. All bboxes are long rectangle:
> > > ![image](https://user-images.githubusercontent.com/12689630/78970409-1309b680-7abe-11ea-9ac7-d2c7047ca95c.png)

I think you might be doing it wrong. I just trained coco from scratch for thousands of steps, the anchors seem fine

![f5f1ee78b8fe4b05a50367a606386474](https://user-images.githubusercontent.com/5842006/78973070-b059fe80-7b41-11ea-80a2-290ed9cae8b0.jpg)

**Comment by kaishijeng** [comment]
I didn't change anchor config. See the attachment for my trained weight.
[efficientdet-d0_7_167586.zip](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/files/4460454/efficientdet-d0_7_167586.zip)

**Comment by kaishijeng** [comment]
coco.yml:
project_name: coco  # also the folder name of the dataset that under data_path folder
train_set: train2017
val_set: val2017
num_gpus: 1

# mean and std in RGB order, actually this part should remain unchanged as long as your dataset is similar to coco.
mean: [0.485, 0.456, 0.406]
std: [0.229, 0.224, 0.225]

args in train.py:
def get_args():
    parser = argparse.ArgumentParser('Yet Another EfficientDet Pytorch: SOTA object detection network - Zylo117')
    parser.add_argument('-p', '--project', type=str, default='coco', help='project file that contains parameters')
    parser.add_argument('-c', '--compound_coef', type=int, default=0, help='coefficients of efficientdet')
    parser.add_argument('-n', '--num_workers', type=int, default=12, help='num_workers of dataloader')
    parser.add_argument('--batch_size', type=int, default=12, help='The number of images per batch among all devices')
    parser.add_argument('--head_only', type=bool, default=False,
                        help='whether finetunes only the regressor and the classifier, '
                             'useful in early stage convergence or small/easy dataset')
    parser.add_argument('--lr', type=float, default=1e-4)
    parser.add_argument('--alpha', type=float, default=0.25)
    parser.add_argument('--gamma', type=float, default=1.5)
    parser.add_argument('--num_epochs', type=int, default=500)
    parser.add_argument('--val_interval', type=int, default=1, help='Number of epoches between valing phases')
    parser.add_argument('--save_interval', type=int, default=500, help='Number of epoches between saving')
    parser.add_argument('--es_min_delta', type=float, default=0.0,
                        help='Early stopping\'s parameter: minimum change loss to qualify as an improvement')
    parser.add_argument('--es_patience', type=int, default=0,
                        help='Early stopping\'s parameter: number of epochs with no improvement after which training will b
e stopped. Set to 0 to disable this technique.')
    parser.add_argument('--data_path', type=str, default='datasets/', help='the root folder of dataset')
    parser.add_argument('--log_path', type=str, default='logs/')
    parser.add_argument('--load_weights', type=str, default=None,
                        help='whether to load weights from a checkpoint, set None to initialize, set \'last\' to load last
checkpoint')
    parser.add_argument('--saved_path', type=str, default='logs/')

Training script:
python train.py -c 0 --batch_size 12 --data ./data/

**Comment by kaishijeng** [comment]
If I use your pretrained d0 weight, the result looks OK. So test program should be OK

![image](https://user-images.githubusercontent.com/12689630/78973792-721ef980-7ac5-11ea-9952-fa8c29e4e95c.png)

**Comment by zylo117** [comment]
ok, I tried to debug it, meanwhile, you can create a test/ folder and uncomment this line to debug,
You can visualize the predicted boxes there.

```
cls_loss, reg_loss = criterion(classification, regression, anchors, annot,
                                                   # imgs=imgs, obj_list=params.obj_list  # uncomment this to debug
                                                   )
```

**Comment by zylo117** [comment]
> I didn't change anchor config. See the attachment for my trained weight.
> [efficientdet-d0_7_167586.zip](https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/files/4460454/efficientdet-d0_7_167586.zip)

CALM DOWN, I thought there is a bug or what. It turns out ok, your weights are fine. There is only some boxes are long rectangle, not all of them. It quite normal on the early stage.

![img_inferred_d0_this_repo_0](https://user-images.githubusercontent.com/5842006/78974227-1a73a300-7b44-11ea-96d5-19e8ba919ec1.jpg)

**Comment by kaishijeng** [comment]
OK. What is difference if I set force_input_size to None.?

**Comment by zylo117** [comment]
> OK. What is difference if I set force_input_size to None.?

use None, then the input size equals training data input size, or feature map will be lager or smaller, the regressor is not familiar to it, especially early stage weights

**Comment by kaishijeng** [comment]
Question:

What are ls loss and reg loss shown during training?
  >>Cls loss: 0.04985. Reg loss: 0.02022. Total loss: 0.07007

What are values of these 2 losses which we consider training reaches  good state?

**Comment by zylo117** [comment]
loss is just loss, not to mention this is just training loss, you could be suffering overfitting. you need to compare it with your val loss. better yet, you need to evaluate mAP/mAR or something else.

**Comment by hibrahimozturk** [comment]
Hi @zylo117,

Firstly, Thank you for this good pytorch implementation and pretrained weights which you shared.

I want to learn your training strategy. For example how many epochs have you trained the network with adamw optimizer before switching to sgd ? Can you share loss graphs of adamw and sgd stages  ?
